core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -354,8 +356,45 @@ public Response getBucket(@PathParam("bucket") final String bucket,
             if (prefixParam != null) {
               path = parsePathWithDelimiter(path, prefixParam, AlluxioURI.SEPARATOR);
             }
-            ListStatusPOptions options = ListStatusPOptions.newBuilder().setRecursive(true).build();
-            children = userFs.listStatus(new AlluxioURI(path), options);
+            String startAfter = (markerParam != null ? markerParam
+                : (startAfterParam != null ? startAfterParam : ""));
+            ListStatusPartialPOptions options = ListStatusPartialPOptions.newBuilder()
+                .setStartAfter(startAfter)
+                .setBatchSize(maxKeys)
+                .setOptions(ListStatusPOptions.newBuilder().setRecursive(true))
+                .buildPartial();
+
+            ListStatusPartialResult partialResult =
+                userFs.listStatusPartial(new AlluxioURI(path), options);
+            children = partialResult.getListings();
+
+            ListBucketResult bucketResult =
+                new ListBucketResult(bucket, children, listBucketOptions);
+
+            while (bucketResult.getContents().size() < maxKeys && !children.isEmpty()) {
+              options = options.toBuilder()
+                  .setStartAfter(children.get(children.size() - 1).getPath())
+                  .buildPartial();
+              partialResult = userFs.listStatusPartial(new AlluxioURI(path), options);
+              children = partialResult.getListings();
+              ListBucketResult partialBucketResult = bucketResult;
+              bucketResult =  new ListBucketResult(bucket, children, listBucketOptions);
+              if (!partialBucketResult.getContents().isEmpty()) {
+                bucketResult.addContent(partialBucketResult.getContents());
+              }
+            }
+            // check for more results to tell the client whether the response is truncated
+            if (children.size() > 0 && !bucketResult.isTruncated()) {
+              options = options.toBuilder()
+                  .setStartAfter(children.get(children.size() - 1).getPath())
+                  .setBatchSize(1).buildPartial();
+              partialResult = userFs.listStatusPartial(new AlluxioURI(path), options);
```

Comment:
Is there a bug that means you need to get an extra result from the master?

Can you describe the situation where you need this extra call, and maybe I can fix it?

Reply:
This is used to determine whether there is more data, thereby modifying the value of mIsTruncated.